Make $it in filters inside select/expand reference the resource path

KenitoInc opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

*This pull request fixes issue #1837 .*

### Description
Whenever we have an `expand` query, we are adding a `$it` `rangeVariable` whose `navigationSource` is the `resource path entity set` instead of the expanded entity set.
ODATA spec: Example 106 http://docs.oasis-open.org/odata/odata/v4.01/csprd05/part2-url-conventions/odata-v4.01-csprd05-part2-url-conventions.html#sec_it

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

KenitoInc commented:
@kosinsky Have a look at this PR

kosinsky commented:
Looks good for me. May be add a few more complex tests with deeper expands

KenitoInc commented:
> Looks good for me. May be add a few more complex tests with deeper expands

@kosinsky Done

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -274,22 +274,22 @@ private SelectItem GenerateSelectItem(SelectTermToken tokenIn)
 
             // $compute
-            ComputeClause compute = BindCompute(tokenIn.ComputeOption, targetNavigationSource, elementTypeReference);
+            ComputeClause compute = BindCompute(tokenIn.ComputeOption, null, targetNavigationSource, elementTypeReference);
```
1) I can't understand why input "null" in $select.
2) what's the difference between "navigationSource" and "TargetNavigationSource"?

xuzhg replied:
And complex select can also use the "this.NavigationSource"? why input "null"?

---
In reply to: [515430616](https://github.com/OData/odata.net/pull/1906#discussion_r515430616) [](ancestors = 515430616)

KenitoInc replied:
@xuzhg 
Responses below:
1. We only want to add another `$it` Range variable to the stack for the `ExpandTermToken`, not the `SelectTermToken`.
2. `NavigationSource` is the parent entity set, `TargetNavigationSource` is the expanded entity set.

KenitoInc replied:
@xuzhg I added a fix to `Select`, so we are adding a `$it` RangeVariable for the `SelectTermToken`

KenitoInc replied:
#Resolved

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -820,20 +841,22 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
+            bool collapsed = false)
         {
-            if (targetNavigationSource == null && elementType == null)
+            if (navigationSource == null && targetNavigationSource == null && elementType == null)
```
I think adding `navigationSource == null` to this condition makes it possible to get a null-reference exception online 868.

With the original if statement condition `(targetNavigationSource == null && elementType == null)`   if `elementType` is null, then we can guarantee that `targetNavigationSource` is non-null on line 845, otherwise the method would have terminated on line 861.

But with the new condition, it's possible that both `targetNavigationSource` and `elementType` are `null` on line 868 (if `navigationSource` is non-null), leading to a potential null-reference exception on `targetNavigationSource.EntityType()`

KenitoInc replied:
#Resolved

mikepizzo replied:
Good catch!

---
In reply to: [530665718](https://github.com/OData/odata.net/pull/1906#discussion_r530665718) [](ancestors = 530665718)

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/RangeVariableBinder.cs:
```diff
@@ -24,7 +24,11 @@ internal static SingleValueNode BindRangeVariableToken(RangeVariableToken rangeV
+            // We are using FirstOrDefault instead of SingleOrDefault since we can have >1 $it rangeVariable in the Stack.
+            // If we added another $it rangeVariable to the Stack (when binding the expandToken or selectToken),
+            // We create the RangeVariableReferenceNode using the added $it rangeVariable.
+            // ASSUMPTION: Max 2 $it rangeVariables can be added to the Stack.
```
The method is called for all rangeVariables. 
Initially, a `QueryToken` could only have one `rangeVariable`. Since we have added another `$it` rangeVariable in `SelectTermToken` and `ExpandTermToken`, we can have a max of 2 $it rangeVariables in the stack.

KenitoInc replied:
I refractored the code to have only 1 $it range variable.
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -378,30 +399,30 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
 
             // $compute
-            ComputeClause computeOption = BindCompute(tokenIn.ComputeOption, targetNavigationSource);
+            ComputeClause computeOption = BindCompute(tokenIn.ComputeOption, this.NavigationSource, targetNavigationSource);
```
Actually it should be `this.ResourcePathNavigationSource`.
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -416,11 +437,11 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
         /// <param name="navigationSource">The navigation source.</param>
         /// <returns>The null or the built apply clause.</returns>
-        private ApplyClause BindApply(IEnumerable<QueryToken> applyToken, IEdmNavigationSource navigationSource)
+        private ApplyClause BindApply(IEnumerable<QueryToken> applyToken, IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource)
```
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -435,11 +456,11 @@ private ApplyClause BindApply(IEnumerable<QueryToken> applyToken, IEdmNavigation
         /// <param name="elementType">The target element type.</param>
         /// <returns>The null or the built compute clause.</returns>
-        private ComputeClause BindCompute(ComputeToken computeToken, IEdmNavigationSource navigationSource, IEdmTypeReference elementType = null)
+        private ComputeClause BindCompute(ComputeToken computeToken, IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource, IEdmTypeReference elementType = null)
```
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -456,12 +477,12 @@ private ComputeClause BindCompute(ComputeToken computeToken, IEdmNavigationSourc
         /// <param name="collapsed">The collapsed boolean value.</param>
         /// <returns>The null or the built filter clause.</returns>
-        private FilterClause BindFilter(QueryToken filterToken, IEdmNavigationSource navigationSource,
+        private FilterClause BindFilter(QueryToken filterToken, IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource,
```
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -499,11 +520,11 @@ private ComputeClause BindCompute(ComputeToken computeToken, IEdmNavigationSourc
         /// <param name="elementType">The Edm element type.</param>
         /// <returns>The null or the built search clause.</returns>
-        private SearchClause BindSearch(QueryToken searchToken, IEdmNavigationSource navigationSource, IEdmTypeReference elementType)
+        private SearchClause BindSearch(QueryToken searchToken, IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource, IEdmTypeReference elementType)
```
#Resolved

KenitoInc commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -523,12 +544,12 @@ private SearchClause BindSearch(QueryToken searchToken, IEdmNavigationSource nav
         /// <returns>The null or the built select and expand clause.</returns>
         private SelectExpandClause BindSelectExpand(ExpandToken expandToken, SelectToken selectToken,
-            IList<ODataPathSegment> segments, IEdmNavigationSource navigationSource, IEdmTypeReference elementType,
+            IList<ODataPathSegment> segments, IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource, IEdmTypeReference elementType,
```
#Resolved

mikepizzo commented on src/Microsoft.OData.Core/UriParser/Binders/RangeVariableBinder.cs:
```diff
@@ -24,7 +24,11 @@ internal static SingleValueNode BindRangeVariableToken(RangeVariableToken rangeV
+            // If we added another $it rangeVariable to the Stack (when binding the expandToken or selectToken),
+            // We create the RangeVariableReferenceNode using the added $it rangeVariable.
+            // ASSUMPTION: Max 2 $it rangeVariables can be added to the Stack.
+            RangeVariable rangeVariable = state.RangeVariables.FirstOrDefault(p => p.Name == rangeVariableToken.Name);
```
>FirstOrDefault [](start = 63, length = 14)

I am also confused by this logic.  If there are two $it range variables, how do we know that the desired range variable always the first?

KenitoInc replied:
@mikepizzo  The first implicit rangeVariable is pushed to the stack in line 877. The one we are adding is pushed in line 889. The one on top of the stack will be the desired one.

KenitoInc commented:
Hey @mikepizzo 
I have updated the PR as we had discussed.
-	Removed the code that added the implicit Range Variable to the Stack. So we will only have the explicit range variable in the stack.
-	Reverted the code in the RangeVariableBinder from FirstOrDefault to SingleOrDefault
-	Added the comments to explain the different between the implicit and explicit range variables.

mikepizzo reviewed (approved):
:shipit:

mikepizzo commented:
LGTM -- thanks Kennedy!

---
In reply to: [741140266](https://github.com/OData/odata.net/pull/1906#issuecomment-741140266) [](ancestors = 741140266)

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/BindingState.cs:
```diff
@@ -151,6 +151,11 @@ internal List<CustomQueryOptionToken> QueryOptions
+        /// <summary>
+        /// The navigation source at the root.
+        /// </summary>
+        internal IEdmNavigationSource ResourcePathNavigationSource { get; set; }
```
>ResourcePathNavigationSource [](start = 38, length = 28)

maybe name it as "RootNavigationSource"?

KenitoInc replied:
`ResourcePath` is more appropriate. Check here https://docs.microsoft.com/en-us/odata/concepts/url-components

gathogojr replied:
The docstring says "The navigation source at the root" so `RootNavigationSource` sould also be a good fit if not better...

KenitoInc replied:
Let me update the docstring. Should be `Navigation source at the resource path`

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -34,6 +33,11 @@ internal sealed class SelectExpandBinder
+        /// <summary>
+        /// The navigation source at the resource path.
+        /// </summary>
+        private readonly IEdmNavigationSource resourcePathNavigationSource;
```
>resourcePathNavigationSource [](start = 46, length = 28)

can the root navigation source figure out from "ODataUriParserConfiguration" or other place?

KenitoInc replied:
We are setting it from the `ODataPathInfo`. Having the property in `SelectExpandBinder` allows us to pass it to Bind Methods since it is not stored in the `SelectTermToken` and `ExpandTermToken`

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -820,34 +847,53 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
 
         private static MetadataBinder BuildNewMetadataBinder(ODataUriParserConfiguration config,
+            IEdmNavigationSource navigationSource,
             IEdmNavigationSource targetNavigationSource,
```
can you add a comment for this method to illustrate the difference between "navigationSource" and "targetNavigationSource"?

KenitoInc replied:
I have renamed `navigationSource` to `resourcePathNavigationSource` which makes it clear

xuzhg commented on test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/SelectExpandFunctionalTests.cs:
```diff
@@ -1273,6 +1273,71 @@ public void SelectAndExpandShouldWorkOnSelectOpenProperty()
+
+            IEdmStructuredType right = (((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Right as SingleValuePropertyAccessNode).Property.DeclaringType;
+            IEdmStructuredType left = (((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Left as SingleValuePropertyAccessNode).Property.DeclaringType;
+            Assert.Equal(expectedRight, right);
```
It's too long to maintain and read. Better to separate multiple lines?

KenitoInc replied:
Done.
#Resolved

xuzhg reviewed (approved):
:shipit:

gathogojr commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -820,34 +847,53 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
+                // This $it rangeVariable will be added the Stack.
+                // We are adding a rangeVariable whose navigationSource is the resource path entity set.
+                // ODATA spec: Example 106 http://docs.oasis-open.org/odata/odata/v4.01/csprd05/part2-url-conventions/odata-v4.01-csprd05-part2-url-conventions.html#sec_it
+                RangeVariable explicitRangeVariable = NodeFactory.CreateImplicitRangeVariable(
```
We call `NodeFactory.CreateImplicitRangeVariable()` here and assign it to a variable named `explicitRangeVariable`... Is use of "explicit" in the variable name deliberate?

KenitoInc replied:
It's deliberate --
We use the same Factory method to create both the implicit and explicit range variables.

gathogojr commented on src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs:
```diff
@@ -820,34 +847,53 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
+            // For example if we have https://url/Books?$expand=Authors($filter=Name eq $it/Name)
+            // $filter=Name will reference Authors(the expanded entity).
+            // $it/Name will reference Books(the resource identified by the path).
+            // The BindingState ImplicitRangeVariable property will store the $it that references the expanded/selected item (The Implicit Range Variable).
```
I have to admit, I'm not sure I understand the two statements on Line 873 and 874...

gathogojr reviewed (approved):
Left a few comments otherwise LGTM
